cmd/sriov/main.go

```diff
@@ -270,6 +270,29 @@ func cmdDel(args *skel.CmdArgs) error {
 
 	sm := sriov.NewSriovManager()
 
+	netns, err := ns.GetNS(args.Netns)
+	if err != nil {
+		// according to:
+		// https://github.com/kubernetes/kubernetes/issues/43014#issuecomment-287164444
+		// if provided path does not exist (e.x. when node was restarted)
+		// plugin should silently return with success after releasing
+		// IPAM resources
+		_, ok := err.(ns.NSPathNotExistErr)
+		if ok {
+			logging.Debug("Exiting as the network namespace does not exists anymore",
+				"func", "cmdDel",
+				"netConf.DeviceID", netConf.DeviceID,
+				"args.Netns", args.Netns)
+			return nil
```

Comment:
in this case dont we want to skip to L315 and mark pci address as released ?

Reply:
I think it's better not:

The scenario I'm trying to solve here is when a CmdDelete is called when the device has already been allocated to another Pod. If we call `DeleteAllocatedPCI(...)`, we are freeing up a device on a running pod.

WDYT?

Reply:
can this really happen ? we use the pci allocator to check if the device is allocated in CmdAdd, see [1]

so even if NS was deleted and device was re-assigned to a new pod, its CNI call would fail on ADD until we released the device.

[1] https://github.com/k8snetworkplumbingwg/sriov-cni/blob/fca6591297c0e52b6522573bd367ccb7d6551fd0/pkg/config/config.go#L60

Reply:
You're right; in the scenario I just described, it probably can't happen.
What we saw happening is related to a very long execution of [`ipam.ExecDel(...)`](https://github.com/k8snetworkplumbingwg/sriov-cni/pull/313/files#diff-30bddd0b0a3414968a5fa92d19e50ad3168e273728fd52ba867afbc442defe62R255).
This scenario includes 3 sriov-cni calls:
a. **first_cniDEL** starts and call ipam.ExecDel(...). this is going to take a very long time (~2 minutes)
b. **second_cniDEL** is invoked and, for some reason, it completes quite quickly (ipam deletion, vlan/mac reset, ...). PIC is deallocated, netns is deleted.
c. **cniADD** is  invoked. It assigns the VF to a new pod, configuring VLAN, MAC, ...
d. **first_cniDEL** continues its execution, as the `ipam.ExecDel` finished. Here, what I want to avoid is resetting the VF MAC, VLAN and so on.

Reply:
maybe a better solution is to use lock file to handle on a single cmd ADD/DEL at a time for a given ns, network, interface ?
that way second cni DEL will fail and prevent freeing up the PCI in the allocator ?

Reply:
I was thinking the same but I don't want to have another file maybe we can use the existing PCI file and have the lock on that file?

Reply:
we can flock the pci file if it works. just need to make sure we release when we exit ADD and DEL cmds.

Reply:
looking at the code, we save PCI device at the end of the cmd add call.
in this case we want to "lock" the call for the ns/network/interface at the beginning.
so if we want to use the same file we need to be careful (save at the beginning and release in case of failure)

Reply:
you are right so maybe having a tmp file for `ns/network/interface` is the right way going forward but we must be sure we clean that file.

and we cover a case of unexpected reboot so maybe putting this lock file in /tmp/ so on reboot it will get clean?

Reply:
sounds good.